drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -300,44 +301,148 @@ impl Transceivers {
+    pub fn enable_power<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_POWER_EN0)
+    }
+
+    /// Clear power enable bits per the specified `mask`. Controls whether or
+    /// not a module's hot swap control will be turned on by the FPGA upon
+    /// module insertion.
+    pub fn disable_power<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_POWER_EN0)
+    }
+
+    /// Set ResetL bits per the specified `mask`. This directly controls the
+    /// ResetL signal to the module.
+    pub fn deassert_reset<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_MOD_RESETL0)
+    }
+
+    /// Clear ResetL bits per the specified `mask`. This directly controls the
+    /// ResetL signal to the module.
+    pub fn assert_reset<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_MOD_RESETL0)
+    }
+
+    /// Set LpMode bits per the specified `mask`. This directly controls the
+    /// LpMode signal to the module.
+    pub fn assert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Clear LpMode bits per the specified `mask`. This directly controls the
+    /// LpMode signal to the module.
+    pub fn deassert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Have the SP execute a reset by manually writing the ResetL bits in the
+    /// FPGA. SFF-8679 states that the reset pulse needs to be 10 microseconds
+    /// (t_reset_init), but the module will not be ready for normal operation
+    /// until 2 seconds (t_reset) after reset is released.
+    pub fn module_reset<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.assert_reset(mask)?;
+        userlib::hl::sleep_for(1);
+        self.deassert_reset(mask)?;
+        userlib::hl::sleep_for(2000);
+        Ok(())
+    }
+
+    /// Sequence of actions to turn off modules per the specified `mask
+    pub fn power_mode_off<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.disable_power(mask)?;
+        // lpmode is being deasserted while the module is off to
+        // because it will leak into the main power rail if
+        // actively driven in the absence of power for at least
+        // some modules.
+        self.deassert_lpmode(mask)?;
+        self.assert_reset(mask)?;
```

Comment:
What happens if one of these operations fails and we return early with the `Err`? Is there a danger of leaving the FPGA in an inconsistent state?